### PR TITLE
Texture and buffer bindings polyfill fixes

### PR DESCRIPTION
**Pull request by ddenisyuk**

### CODE REVIEW

File: src/worker-thread/canvas/WebGLRenderingContextPolyfill.ts [hunk 1]
```diff
@@ -1968,73 +2173,83 @@ export class WebGLRenderingContextPolyfill extends GLConstants implements WebGL2
...
-        break;
-      }
-      case this.UNIFORM_BUFFER: {
-        this._buffers.uniformBuffer = buffer;
-        break;
-      }
-      case this.PIXEL_PACK_BUFFER: {
-        this._buffers.pixelPackBuffer = buffer;
-        break;
+  _webglError(errorCode: GLenum, errorName: string, operation: string, message: string): void {
```
Comment (olitvinenko) [inline src/worker-thread/canvas/WebGLRenderingContextPolyfill.ts]: explicit access modifier is missed. the default one in TS is public.

File: src/worker-thread/canvas/WebGLRenderingContextPolyfill.ts [hunk 2]
```diff
@@ -1968,73 +2173,83 @@ export class WebGLRenderingContextPolyfill extends GLConstants implements WebGL2
...
+    this._error = errorCode;
+    console.warn(`WebGL: ${errorName}(${errorCode}): ${operation}: ${message}`);
+  }
+
+  _validateBinding(candidate: TransferrableGLObject | null, validator: Function, operation: string, expectedType: string, index: number): boolean {
```
Comment (olitvinenko) [inline src/worker-thread/canvas/WebGLRenderingContextPolyfill.ts]: explicit access modifier is missed. the default one in TS is public.

File: src/worker-thread/canvas/WebGLRenderingContextPolyfill.ts [hunk 3]
```diff
@@ -1968,73 +2173,83 @@ export class WebGLRenderingContextPolyfill extends GLConstants implements WebGL2
...
-        return this._buffers.drawFramebuffer;
       case this.READ_FRAMEBUFFER:
-        return this._buffers.readFramebuffer;
+        return this._buffers[target] || null;
       default:
         throw new Error(`Unexpected target: ${target}`);
     }
   }
+
+  _bindBuffer(target: GLenum, buffer: vGLBuffer | null) {
```
Comment (olitvinenko) [inline src/worker-thread/canvas/WebGLRenderingContextPolyfill.ts]: explicit access modifier is missed

File: src/worker-thread/canvas/WebGLRenderingContextPolyfill.ts [hunk 4]
```diff
@@ -1968,73 +2173,83 @@ export class WebGLRenderingContextPolyfill extends GLConstants implements WebGL2
...
+      }
+    }
+
+    this._buffers[target] = buffer;
+    if (buffer != null && !buffer.bindings.includes(target)) {
+      buffer.bindings.push(target);
+    }
+  }
+
+  _deleteBuffer(buffer: vGLBuffer): void {
```
Comment (olitvinenko) [inline src/worker-thread/canvas/WebGLRenderingContextPolyfill.ts]: explicit access modifier is missed

File: src/worker-thread/canvas/WebGLRenderingContextPolyfill.ts [hunk 5]
```diff
@@ -1968,73 +2173,83 @@ export class WebGLRenderingContextPolyfill extends GLConstants implements WebGL2
...
-        return this._buffers.pixelUnpackBuffer;
       case this.FRAMEBUFFER:
-        return this._buffers.framebuffer;
       case this.RENDERBUFFER:
-        return this._buffers.renderbuffer;
       case this.DRAW_FRAMEBUFFER:
-        return this._buffers.drawFramebuffer;
       case this.READ_FRAMEBUFFER:
-        return this._buffers.readFramebuffer;
+        return this._buffers[target] || null;
```
Comment (olitvinenko) [inline src/worker-thread/canvas/WebGLRenderingContextPolyfill.ts]: this._buffers[target] ?? null

File: src/worker-thread/canvas/WebGLRenderingContextPolyfill.ts [hunk 6]
```diff
@@ -204,89 +215,240 @@ export class WebGLRenderingContextPolyfill extends GLConstants implements WebGL2
...
+      case this.TEXTURE_2D_ARRAY:
+      case this.TEXTURE_3D: {
+        if (texture != null) {
+          if (texture.boundTarget != null && texture.boundTarget != target) {
+            // WebGL: INVALID_OPERATION: bindTexture: textures can not be used with multiple targets
+            this._webglError(this.INVALID_OPERATION, 'INVALID_OPERATION', 'bindTexture', 'textures can not be used with multiple targets');
+            return;
+          }
+          texture.boundTarget = target;
+          if (!texture.texturesUnits.includes(this._activeTexture)) {
```
Comment (olitvinenko) [inline src/worker-thread/canvas/WebGLRenderingContextPolyfill.ts]: as far as I understood you have 2 caches for texture units:
1 - vGLTexture.texturesUnits
2 - WebGLRenderingContextPolyfill._boundTextures

but, I guess, you can use WebGLRenderingContextPolyfill._boundTextures only.

```
for (const textureUnit in this._boundTextures) {
    const target2Texture = this._boundTextures[textureUnit];
    for (const target in target2Texture) {
        const texture = target2Texture[target];
        //
        
        // to delete texture binding
        delete target2Texture[target];
    }

    // to delete unit
    delete _boundTextures[textureUnit];
}
```

Comment (olitvinenko) [reply src/worker-thread/canvas/WebGLRenderingContextPolyfill.ts]: pretty much the same approach could be used for vGLBuffer.bindings as well

**Review by olitvinenko — APPROVED** [review]

File: src/worker-thread/canvas/WebGLRenderingContextPolyfill.ts (hunk 6 above)
Comment (ddenisyuk) [reply src/worker-thread/canvas/WebGLRenderingContextPolyfill.ts]: done, pls review

File: src/worker-thread/canvas/WebGLRenderingContextPolyfill.ts (hunk 5 above)
Comment (ddenisyuk) [reply src/worker-thread/canvas/WebGLRenderingContextPolyfill.ts]: done

File: src/worker-thread/canvas/WebGLRenderingContextPolyfill.ts (hunk 4 above)
Comment (ddenisyuk) [reply src/worker-thread/canvas/WebGLRenderingContextPolyfill.ts]: done

File: src/worker-thread/canvas/WebGLRenderingContextPolyfill.ts (hunk 3 above)
Comment (ddenisyuk) [reply src/worker-thread/canvas/WebGLRenderingContextPolyfill.ts]: done

File: src/worker-thread/canvas/WebGLRenderingContextPolyfill.ts (hunk 2 above)
Comment (ddenisyuk) [reply src/worker-thread/canvas/WebGLRenderingContextPolyfill.ts]: done

File: src/worker-thread/canvas/WebGLRenderingContextPolyfill.ts (hunk 1 above)
Comment (ddenisyuk) [reply src/worker-thread/canvas/WebGLRenderingContextPolyfill.ts]: done

**Review by olitvinenko — APPROVED** [review]

**Pull request merged**